### ISSUE TITLE
Incorrect aspect ratio for thumbnail/preview

### ISSUE THREAD

**Issue by cbhushan** [bpatrik/pigallery2#212]
**Describe the bug**
Thumbnail/preview has incorrect aspect ratio for one of my images, as shown in screenshot below. The image is also linked below. 

**Photo/video (optional) that causes the bug**
See zip: https://www.dropbox.com/s/ee7vw513vl79nuv/testimg.tar.gz?dl=0 

**Screenshots**
Left thumbnail has correct aspect ratio. Right one (__a20180218_143832_dc8f44e06e4a.jpg) is incorrect. Both images are attached in link above.
![image](https://user-images.githubusercontent.com/3436053/103162601-9be98f00-47c0-11eb-9559-2193fb9d21f9.png)

**Environment (please complete the following information):**
 - Browser: Firefox 84.0 (64-bit)
 - OS: Ubuntu 20.04.1
 
**Used app version**:
- docker-nightly. See screenshot below for version.
![image](https://user-images.githubusercontent.com/3436053/103162679-832da900-47c1-11eb-93c2-54f88f2d5fca.png)

docker-compose.yml:
```
version: '3'
services:
  pigallery-db:
    container_name: pigallery-db
    image: mariadb
    volumes:
      - db-data:/var/lib/mysql
    environment:
      - MYSQL_ROOT_PASSWORD=putsomethinginherethatyouwouldlikeasthepassword
      - MYSQL_USER=pigallery2
      - MYSQL_PASSWORD=pigallery2_pass
      - MYSQL_DATABASE=pigallery2

  pigallery2:
    image: bpatrik/pigallery2:nightly-stretch
    command: sh -c 'bin/wait-for pigallery-db:3306 -- --Server-Database-mysql-host=pigallery-db --Server-Database-mysql-username=pigallery2 --Server-Database-mysql-password=pigallery2_pass --Server-Database-mysql-database=pigallery2'
    container_name: pigallery2
    environment:
      - NODE_ENV=production
    volumes:
      - "/mnt/data/tmp/PiGallery2/config:/app/data/config" # CHANGE ME
      - "db-data:/app/data/db"
      - "/mnt/data/tmp/pic_collection:/app/data/images/pic_collection" # CHANGE ME
      - "/mnt/data/tmp/pic_collection2:/app/data/images/pic_collection2" # CHANGE ME
      - "/mnt/data/tmp/PiGallery2/tmp:/app/data/tmp" # CHANGE ME
    ports:
      - 80:80
    restart: always

volumes:
  db-data:

```

**Comment by cbhushan** [comment]
One of the images were modified in by picasa desktop in some ways. I think `__a20180218_143832_dc8f44e06e4a.jpg` is the original image, but I may be wrong.

BTW, thanks for this piece of software! :+1:  It seems to align well with my needs. I wish timeline and tag adding features were also added :smile: (I have read other issues related to these).

**Comment by cbhushan** [comment]
Looks like identical issue as #184. Feel free to close this. I also tested with chromium browser and the thumbnail have same issue. But folder art are different. (the folder `__a` has `__a20180218_143832_dc8f44e06e4a.jpg`)

![image](https://user-images.githubusercontent.com/3436053/103162902-bfaed400-47c4-11eb-9500-73c7b4ed07f7.png)
![image](https://user-images.githubusercontent.com/3436053/103162911-d6edc180-47c4-11eb-896a-5b1c2e5ab7ab.png)

**Comment by cbhushan** [comment]
BTW, the issue is not limited to this one image. I have folder full of images with this issue, but cant share them as they are more personal in nature.

**Comment by snoopyCode** [comment]
This could be a mismatch in the images between exif orientation and actual rotation info in the image. If you e.g. check the exif data with `exiftool` you may find that width and height don't match what you expect because the orientation is something else than `1`. pi-gallery does not seem to take care of this while loading images.

I corrected my images to overcome the issue ([Here is the script for that](https://github.com/snoopyCode/picture-tools.git))

**Comment by edipdincer** [comment]
Have same issue. All vertical oriented images' thumbnails are distorted. I tried Nightly Docker, but still...

**Comment by h0n2a** [comment]
I have the same issue too.
What I don't understand is, why this issue affects thumbnails only. If I open the photo, it's displayed correctly.

Btw., tried to change some css parameters in the browser dev console, and if I set "width:auto;" for <app-gallery-grid-photo> (instead of the current computed value), the result si good, thumbnails are displayed correctly.
Maybe this could be a workaround until you resolve the issue finaly?
Tested with firefox and chrome.

**Comment by vomov** [comment]
> Btw., tried to change some css parameters in the browser dev console, and if I set "width:auto;" for `<app-gallery-grid-photo>` (instead of the current computed value), the result si good, thumbnails are displayed correctly.
Can confirm this appears to solve the issue. It may have some consequences for the layout, but those appear minor.

**Comment by bseclier** [comment]
Same problem here, since the 1.8.2 version (no problem with the 1.8.0 version).

**Comment by bpatrik** [comment]
I replied at: https://github.com/bpatrik/pigallery2/issues/234#issuecomment-798933795

**Comment by bpatrik** [comment]
I did some digging and it is not simple.

I was testing with three photos:
`gold.jpg`, `roof.jpg` and `canyon.jpg`. All photos are portrait.

The relevant part of the exif data:
roof:
```
{
  size: { width: 2304, height: 3072 },
  orientation: 8, // Left side bottom
  cameraData: {
    model: 'DMC-LZ7',
    make: 'Panasonic',
  }
}

```
gold:
```
{
  size: { width: 4032, height: 3024 },
  orientation: 6, // Right side top
  cameraData: {
    model: 'Pixel 2',
    make: 'Google',
  },
}

```
canyon
```
{
  size: { width: 4032, height: 3024 },
  orientation: 6, // Right side top
  cameraData: {
    model: 'SM-G955W',
    make: 'samsung',
  }
}

```
Details about the orientation tag [here](https://www.impulseadventure.com/photo/exif-orientation.html)

As you can see for `gold` and `canyon` width < height (not correct!) and orientation tag suggest, it should be rotated. 
While for `roof` height < width (correct) and orientation tag suggest, it should be rotated. 
Based on this data **there is no way to know if a photo should be rotated or not**
I know it is possible to tell, as other tools (windows, some exif manipulators) can tell the difference, but I could not figure it out.

I tried different exif readers (http://metapicz.com/, http://exif.regex.info/exif.cgi, http://exif-viewer.com/) they all show the same exif value. 

I ended up checking if the orientation suggest the photo to be portrait, but the dimensions say it is landscape, I flip the dimensions: https://github.com/bpatrik/pigallery2/blob/8bd04dda8c5dcdd604ad71f2472229cd7e0af70f/src/backend/model/threading/MetadataLoader.ts#L169-L176

Let me know if you figure out the actual issue or if you have photos that still break with this improvement. 

You can try your luck too. Here are the photos:

| Gold              | roof             | Canyon  |
|:-------------:|:-------------:|:---------:|
| ![gold](https://user-images.githubusercontent.com/7512182/112716428-bda10200-8ee6-11eb-9068-ed75296665f9.jpg)      | ![roof](https://user-images.githubusercontent.com/7512182/112716436-c8f42d80-8ee6-11eb-9725-ab966bef2d4a.jpg) |![canyon](https://user-images.githubusercontent.com/7512182/112716445-d27d9580-8ee6-11eb-98a2-80f3a1cd039f.jpg )|

**Comment by h0n2a** [comment]
Thanks for the effort. Unfortunately, the issue is still the same for me.
![gallery](https://user-images.githubusercontent.com/65588883/112732690-c2dc6c00-8f3b-11eb-99df-8a56be798d2b.png) ![240](https://user-images.githubusercontent.com/65588883/112732700-cf60c480-8f3b-11eb-9d16-d713942c6379.jpg)
The left picture is screenshot from the gallery, the right is the thumbnail downloaded from the gallery.

**Comment by bpatrik** [comment]
Did you pulled the latest Docker container? Also you probably need to reset the DB (forgot to mention), as it caches the dimension and does not recalculate using the new algorithm.

Also can you send me the original photo, so that I can test? The thumbnail is generated by the app and that removes some information from it.

**Comment by h0n2a** [comment]
OK, database reset did the trick. Now it is working! Thanks, seems you fixed the issue (at least for me).

**Comment by cbhushan** [comment]
Thanks for the update. I will check the new code later and let you know how it worked.  

> As you can see for `gold` and `canyon` width < height (not correct!) and orientation tag suggest, it should be rotated.
> While for `roof` height < width (correct) and orientation tag suggest, it should be rotated.

I think your interpretation of exif "width" and "height" is inaccurate (although your final solution might be doing the right thing). Exif width and height only define the number of Pixels in X-dimension and Y-dimension respectively of pixel matrix **_before_** any additional rotation. Look for `ImageWidth` and `ImageLength` in [exif specification](https://www.exif.org/Exif2-2.PDF) ([another ref](https://exiv2.org/tags.html)) . They do not imply final width and height after applying all orientation stuff, and may have  width < height even for portrait mode.

I would also suggest to check following: 
- https://github.com/recurser/exif-orientation-examples : Good repo with examples of all possible types of orientation tags for testing.
- See [exif_transpose](https://pillow.readthedocs.io/en/stable/_modules/PIL/ImageOps.html#exif_transpose) function in python Pillow library, which performs the orientation as per exif.
- Check Phil Harvey's response in this thread about the logic to find landscape images: https://exiftool.org/forum/index.php?topic=7523.0

**Comment by cbhushan** [comment]
Results below with new docker image + db reset. Some orientation cases are definitely fixed, while others are not, specially landscape ones.

![image](https://user-images.githubusercontent.com/3436053/112739311-5f553d00-8f41-11eb-81de-36b68abd685f.png)

**Comment by cbhushan** [comment]
I tried to go through the code but was unable to follow all the variable (I am illiterate in JS/TS :unamused: ). 

If you **only** need to get width/height after all the orientation stuff, then following image from [this blog post](https://www.daveperrett.com/articles/2012/07/28/exif-orientation-handling-is-a-ghetto/) seems to imply that you can flip the dimensions whenever you get exif orientation of 5, 6, 7, 8 (i.e. exif-height become width and like-wise) and you should be good. 

![image](https://user-images.githubusercontent.com/3436053/112739677-35e9e080-8f44-11eb-8eb2-36e4748044b7.png)

**Comment by edipdincer** [comment]
In my case, everything works fine. Thanks for your hard work.

**Comment by bpatrik** [comment]
Thank you @cbhushan for looking into it.
https://github.com/recurser/exif-orientation-examples  is an awesome repo. I added these as tests, so hopefully this wont break again in the future.


I did more digging and found the actual issue:

TL;dr: the lib was reading a wrong orientation value.

long version:
I'm using multiple exif, iptc parers (as I could not find one that can do all at once) to get all the metadata for a photo:

1) fs.statSync for silesize, creation date (if i do not find any better from exif)
https://github.com/bpatrik/pigallery2/blob/36ecbcbd3d5d3e659ddf20add456982dcb960d3d/src/backend/model/threading/MetadataLoader.ts#L107

2) `ts-exif-parser` for most of the exif data:  https://github.com/bpatrik/pigallery2/blob/36ecbcbd3d5d3e659ddf20add456982dcb960d3d/src/backend/model/threading/MetadataLoader.ts#L114
Turns out that it was not reliable reading the `orientation` data. **Apparently both the photo and the embedded thumbnail can both have orientations. It is also possible that their orientation is not the same. In case of `roof.jpg` The thumbnail orientation is 8, while the photo orientation is 0.** This lib was returning 8 not specifying if it belonged to the photo or the thumbnail.

3) `ts-node-iptc` for GPSm keywords, caption related info: https://github.com/bpatrik/pigallery2/blob/36ecbcbd3d5d3e659ddf20add456982dcb960d3d/src/backend/model/threading/MetadataLoader.ts#L177

4) `exifreader` for faces and now for orientation as this lib parses that prioperly https://github.com/bpatrik/pigallery2/blob/36ecbcbd3d5d3e659ddf20add456982dcb960d3d/src/backend/model/threading/MetadataLoader.ts#L206


Answering you question `OrientationTypes` is just an enum (ts style enums are integers):
https://github.com/bpatrik/ts-exif-parser/blob/5aac26aa39824fea5af675671174fc53b996e4d8/lib/ExifData.ts#L3-L12

**Comment by cbhushan** [comment]
Thanks for detailed response. The new update seems to work well for sample images. I will update later if all my albums look okay (i moved few things; need to move files back to RPi). 

I have no expertise in JS but this lib claims to be _"fastest and most versatile JavaScript EXIF reading library'"_ : 
https://mutiny.cz/exifr/

**Comment by bpatrik** [comment]
Thanks.
https://mutiny.cz/exifr/ is missing some feature. I opened a ticket for them. Let see if it gets implemented.
